csrc/device_lower/utils.cpp

```diff
@@ -1013,6 +1016,618 @@ bool predicateAtEnd(ForLoop* loop) {
+// to groups. Note that the traversal path is obtained by running BFS from
+// domain to linear_g, so the direction is flipped with respect to how we
+// propagate from linear_g to domain.
+auto fromGroups(
+    const ValGraph& id_graph,
+    const ExprGroup& eg,
+    Direction direction) {
+  return direction == Direction::Forward ? id_graph.outputGroups(eg)
+                                         : id_graph.inputGroups(eg);
+}
+
+auto toGroups(
+    const ValGraph& id_graph,
+    const ExprGroup& eg,
+    Direction direction) {
+  return direction == Direction::Forward ? id_graph.inputGroups(eg)
+                                         : id_graph.outputGroups(eg);
+}
+
+// Do the propagation from linear_g to domain.
+LinearGroupProjection propagate(
+    const LinearGroupProjection& current,
+    const ValGraph& id_graph,
+    const ExprGroup& eg,
+    Direction direction);
+
+LinearGroupProjection propagate(
+    const ValGroup& current,
+    const ValGraph& id_graph,
+    const ExprGroup& eg,
+    Direction direction) {
+  auto from = fromGroups(id_graph, eg, direction);
+  auto to = toGroups(id_graph, eg, direction);
+  if (from.size() == 1) {
+    // If linear_g is equivalent to from, and from is split into two groups,
+    // then these two groups together represents linear_g.
+    NVF_ERROR(to.size() == 2);
+    NVF_ERROR(from.front() == current);
+    return Composition<LinearGroupProjection>{to.front(), to.back()};
+  } else {
+    // If linear_g is merged with another group, then part of the merged group
+    // represents linear_g.
+    NVF_ERROR(from.size() == 2);
+    NVF_ERROR(to.size() == 1);
+    NVF_ERROR(from.front() == current || from.back() == current);
+    return PartOf<LinearGroupProjection>{
+        std::make_shared<LinearGroupProjection>(to.front()),
+        /*inner_extent=*/from.front() == current
+            ? from.back()->front()->as<IterDomain>()->extent()
+            : nullptr,
+        /*selected_extent=*/
+        simplifyExpr(current->front()->as<IterDomain>()->extent())};
+  }
+}
+
+LinearGroupProjection propagate(
+    const PartOf<LinearGroupProjection>& current,
+    const ValGraph& id_graph,
+    const ExprGroup& eg,
+    Direction direction) {
+  auto from = fromGroups(id_graph, eg, direction);
+  auto to = toGroups(id_graph, eg, direction);
+  if (from.size() == 1) {
+    NVF_ERROR(to.size() == 2);
+    NVF_ERROR(related(*current.group, from.front()));
+    auto group = propagate(*current.group, id_graph, eg, direction);
+    if (!group.hasValue()) {
+      return {};
+    }
+    // The result is actually just the
+    //   PartOf{group, current.inner_extent, current.selected_extent}
+    // However, we want to simplify the result. For example, if the inner_extent
+    // is a multiple of the extent of the last item in group, we can simplify
+    // the inner_extent and group by canceling the last item in group.
+    Val* new_inner_extent = current.inner_extent;
```

Comment:
This `new_inner_extent` represents the inner extent of `linear_g` in `group`, right? Why doesn't the inner extent change? For example, if the expr is a merge, and `current.group` corresponds to the outer of the merge, the new group would be the output of the merge, and the inner extent of `linear_g` in the output domain seems to be `current.inner_extent * merge_inner->extent()`. What am I missing?

Reply:
Hmm, it isn't just this assignment, but the comment above also says:

```
 // The result is actually just the
 //   PartOf{group, current.inner_extent, current.selected_extent}
```

I assume `group` here refers to the result of propagation to `*current.group` at line 1303. Per the comment, I assume the rest of the code in this function is not required for correctness, so it seems to indicate `new_inner_extent` is indeed `current.inner_extent`. Please let me know where I misunderstood.

Reply:
Oh, I see. I think I understand your question now. 

> For example, if the expr is a merge, and `current.group` corresponds to the outer of the merge, the new group would be the output of the merge, and the inner extent of `linear_g` in the output domain seems to be `current.inner_extent * merge_inner->extent()`.

This logic is inside the `if (from.size() == 1) {` clause, so if the expr is a merge, then it must be a backward propagation of merge. So `current.group` must be the output of the merge. After propagation, the new group will be a composition of the outer and inner of the merge.

> Per the comment, I assume the rest of the code in this function is not required for correctness, so it seems to indicate `new_inner_extent` is indeed `current.inner_extent`.

Oh, yes, that's right. For example, before the propagation, we have `group = 6` and `inner_extent = 3`. Assume expr is 6 = merge(2, 3), then after propagation, at this point, we will just have "part of the composition of [2, 3], where on the right of linear_g, there is a 3", which can be simplified as "2" later, but for now, we just have the unsimplified case.